### DISCUSSION TITLE
How do you automate MessageBox dialogs in tests?

### DISCUSSION THREAD

**Discussion by mdonatas** [gitextensions/gitextensions#10251] [Q&A] [answered]
I am writing a test where a command produces a dialog, how can I press the button on it from the test? Are there any examples? I couldn't find one.

![image](https://user-images.githubusercontent.com/483659/195682531-d80f186c-8641-49f7-8538-40cf3fea5856.png)

**Comment by mstv** [comment]
The long-term plan is to use `MessageBoxes` for all pop-ups, switch it to `TaskDialog` - and to mock `TaskDialog` (or `MessageBoxes`) for tests, by means of MEF f.i.

**Reply by vbjay** [reply]
You can do this to at least test your things locally

https://github.com/gitextensions/gitextensions/issues/10242#issuecomment-1279391292

Instead of running git bisect commands, you can run the script and it will build and run the program and ask you if it behaved well.  You can use $LASTERRORCODE to help automate the result testing.

**Reply by vbjay** [reply]
https://github.com/gitextensions/gitextensions/pull/10273 Went through and polished scripts

**Comment by RussKie** [comment, marked as answer]
It's not possible to programmatically close a messagebox (without scanning
all windows ans searching for it). TaskDialog was built with this
limitation in mind, and it's possible to close it from code.

**Reply by RussKie** [reply]
Actually, there is a rather clever way to dismiss a modal dialog like so:
```cs
            protected override void WndProc(ref Message m)
            {
                if (m.MsgInternal == User32.WM.ENTERIDLE && m.WParamInternal == (uint)MSGF.DIALOGBOX)
                {
                    HWND dialogHandle = (HWND)m.LParamInternal;
                    PInvoke.PostMessage(dialogHandle, User32.WM.CLOSE);
                }

                base.WndProc(ref m);
            }
```
Though it implies that the test can override `WmdProc`.

For more details refer to https://github.com/dotnet/winforms/pull/8115.
